src/bsoncxx/include/bsoncxx/v_noabi/bsoncxx/stdx/optional.hpp

```diff
@@ -14,93 +14,677 @@
 
-// TODO(MSVC): This would be better expressed as constexpr, but VS2015U1 can't do it.
-const nullopt_t nullopt{::boost::none};
-using ::boost::make_optional;
+template <typename T, typename Ucvr, typename U>
+struct enable_opt_conversion
+    : bsoncxx::detail::conjunction<  //
+          std::is_constructible<T, Ucvr>,
+          bsoncxx::detail::disjunction<  //
+              std::is_same<T, bool>,
+              bsoncxx::detail::negation<
+                  bsoncxx::detail::conjunction<std::is_constructible<T, optional<U>&>,
+                                               std::is_constructible<T, optional<U> const&>,
+                                               std::is_constructible<T, optional<U>&&>,
+                                               std::is_constructible<T, optional<U> const&&>,
+                                               std::is_convertible<optional<U>&, T>,
+                                               std::is_convertible<optional<U> const&, T>,
+                                               std::is_convertible<optional<U>&&, T>,
+                                               std::is_convertible<optional<U> const&&, T>>>>> {};
 
-}  // namespace stdx
-}  // namespace v_noabi
-}  // namespace bsoncxx
+template <typename From, typename To>
+struct enable_opt_value_conversion   //
+    : bsoncxx::detail::conjunction<  //
+          std::is_constructible<To, From&&>,
+          bsoncxx::detail::negation<bsoncxx::detail::is_alike<From, in_place_t>>,
+          bsoncxx::detail::disjunction<
+              bsoncxx::detail::negation<bsoncxx::detail::is_alike<To, bool>>,  //
+              detail::not_an_optional<bsoncxx::detail::remove_cvref_t<From>>>> {};
 
-#elif defined(BSONCXX_POLY_USE_STD_EXPERIMENTAL)
+}  // namespace detail
 
-#include <experimental/optional>
+template <typename T>
+class optional : bsoncxx::detail::equality_operators,
+                 bsoncxx::detail::ordering_operators,
+                 public detail::optional_base_class<T>::type {
+   public:
+    /// The type of value held within this optional
+    using value_type = T;
+    /// An lvalue-reference-to-mutable T
+    using reference = bsoncxx::detail::add_lvalue_reference_t<T>;
+    /// An lvalue-reference-to-const T
+    using const_reference =
+        bsoncxx::detail::add_lvalue_reference_t<bsoncxx::detail::add_const_t<T>>;
+    /// An rvalue-reference-to-mutable T
+    using rvalue_reference = bsoncxx::detail::add_rvalue_reference_t<T>;
+    /// An rvalue-reference-to-const T
+    using const_rvalue_reference =
+        bsoncxx::detail::add_rvalue_reference_t<bsoncxx::detail::add_const_t<T>>;
+    /// A pointer-to-mutable T
+    using pointer = bsoncxx::detail::add_pointer_t<T>;
+    /// A pointer-to-const T
+    using const_pointer = bsoncxx::detail::add_pointer_t<const T>;
 
-namespace bsoncxx {
-namespace v_noabi {
-namespace stdx {
+    // Constructors [1]
+    optional() = default;
+    constexpr optional(nullopt_t) noexcept {}
```

Comment:
`bsoncxx14_constexpr` means that it is only `constexpr` in C++14 mode, since `constexpr` restrictions were relaxed for C++14. Other functions can be `constexpr` in C++11, and have been marked as such.